compiler/rustc_mir_transform/src/inline.rs

```diff
@@ -1119,12 +1123,18 @@ impl<'tcx> MutVisitor<'tcx> for Integrator<'_, 'tcx> {
                     TerminatorKind::Unreachable
                 }
             }
-            TerminatorKind::Resume => {
-                if let Some(tgt) = self.cleanup_block {
-                    terminator.kind = TerminatorKind::Goto { target: tgt }
+            TerminatorKind::Resume => match self.cleanup_block {
+                UnwindAction::Cleanup(tgt) => {
+                    terminator.kind = TerminatorKind::Goto { target: tgt };
                 }
-            }
-            TerminatorKind::Abort => {}
+                UnwindAction::Continue => (),
+                UnwindAction::Unreachable | UnwindAction::Terminate => {
+                    // If the action is terminate, then we would have mapped marked
+                    // all our call-sites as `UnwindAction::Terminate` and no cleanup
+                    // blocks would ever be executed.
+                    terminator.kind = TerminatorKind::Unreachable;
+                }
```

Comment:
I can't quite parse / follow the explanation why we can turn a well-defined `UnwindAction::Terminate` into `TerminatorKind::Unreachable` which has undefined behaviour. The cleanup blocks from the callee might be executed after all.

Reply:
It's actually related to your comment above. I was attempting to turn the inlined call directly into `UnwindAction::Terminate` so that the cleanup block will never be executed.

This is okay because we currently don't guarantee destructors to be called if we know we are going to terminate/abort.

Reply:
Ah, so for `UnwindAction::Terminate` all calls in the callee would be terminated and so cleanup blocks are indeed unreachable.

> This is okay because we currently don't guarantee destructors to be called if we know we are going to terminate/abort.

It would be one thing to leave this unspecified. It is quite another to allow arbitrary nondeterministic choice between either option, which as far as I can see is necessary to explain this transformation. I think we should avoid introducing MIR construct with such surprising semantics.

Reply:
We currently are already skipping over destructors inside nounwind functions (e.g. `extern "C"` functions when `c_unwind` is enabled). I'll raise this question in WG-FFI-unwind.

Reply:
What the initial program behaviour is after MIR building and lowering is not my concern here. The fact that you are changing the behaviour during MIR optimization is. You would need to explain why a program transformation that skips execution of cleanup blocks before terminating is valid. I don't think it should be valid.

Reply:
People in WG-FFI-unwind are inclined to allow this (https://rust-lang.zulipchat.com/#narrow/stream/210922-project-ffi-unwind).

I'll undo the commit that removes `TerminatorKind::Abort` and changes this part to use that terminator instead for now. This optimisation could be added back if a consensus is later reached to allow this sort of transformation.